llvm/lib/Support/Unix/MemoryLinux.h

```diff
@@ -0,0 +1,109 @@
+//===- Unix/MemoryLinux.h - Linux specific Helper Fuctions ------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines Linux specific helper functions for memory management.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIB_SUPPORT_UNIX_MEMORYLINUX_H
+#define LLVM_LIB_SUPPORT_UNIX_MEMORYLINUX_H
+
+#ifndef __linux__
+#error Linux only support header!
+#endif
+
+#include "llvm/Support/Process.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/mman.h>
+#include <sys/syscall.h>
+
+#ifndef MFD_CLOEXEC
+#define MFD_CLOEXEC 0x0001U
+#endif
+#ifndef MFD_EXEC
+#define MFD_EXEC 0x0010U
+#endif
+
+namespace llvm {
+namespace sys {
+namespace {
+
+static inline bool isPermissionError(int err) {
+  // PaX uses EPERM, SELinux uses EACCES
+  return err == EPERM || err == EACCES;
+}
+
+// FIXME: Make this either more low-level C'ish or C++'ish
+static inline bool execProtChangeNeedsNewMapping() {
+  static int status = -1;
+
+  if (status != -1)
+    return status;
+
+  // Try to get the status from /proc/self/status, looking for PaX flags.
+  FILE *f = fopen("/proc/self/status", "re");
```

Comment:
lldb also opens this file, you could follow its lead here:
https://github.com/llvm/llvm-project/blob/main/lldb/source/Host/linux/Support.cpp
https://github.com/llvm/llvm-project/blob/98c73d5df7ff0b5d9c10bc9d44a584d631def1e6/lldb/source/Host/linux/Host.cpp#L178

You should be able to use anything it uses, that's in the `llvm::` namespace.

Reply:
After taking a closer look at `MemoryBuffer` and `StringRef`, I came up with the following which already looks less like C and much more like C++:

```c++
  if (auto file = MemoryBuffer::getFileAsStream("/proc/self/status")) {
    auto pax_flags = (*file)->getBuffer().rsplit("PaX:").second.trim();
    if (!pax_flags.empty())
      // 'M' indicates MPROTECT is enabled
      return (status = pax_flags.find('M') != StringRef::npos);
  }
```

Reply:
I would double check that that splits by line, otherwise you'll end up with `PaX: <rest of file>` and maybe get `M` from somewhere else. Unless trim() is doing a slit at `\n` for you.

Reply:
Actually, I do want `PaX: <rest of the file>` as it is, indeed, the last line in `/proc/self/status`.

Using `trim()` drops the leading whitespace (a `\t`, iirc) and the trailing `\n`, leaving only the flags in `pax_flags`.

Reply:
Oh, and of course, I already tested this works an does the right thing ;)

Reply:
> the last line in /proc/self/status.

It is for now, I appreciate testing that would mean rebuilding the kernel but if it's easy to account for without doing that it's worth it I think.

Perhaps split at the first `\n` after `PaX:`? That handles the end of the file, or the entry if there are future additions.

> I already tested this works an does the right thing ;)

Sure, I assume most people do but sometimes the reasons things work are not the reasons we expect them to, then those change and it not longer works. We just had a bug found parsing `/proc/stat` along these lines.

Reply:
> Perhaps split at the first `\n` after `PaX:`? That handles the end of the file, or the entry if there are future additions.

Easy enough, will do!

> > I already tested this works an does the right thing ;)
> 
> Sure, I assume most people do but sometimes the reasons things work are not the reasons we expect them to, then those change and it no longer works. We just had a bug found parsing `/proc/stat` along these lines.

Parsing strings is hard :/